lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java

```diff
@@ -21,16 +21,35 @@
 import java.util.Collection;
 import java.util.List;
 
+/**
+ * Abstracts a transactional backing store and staging area for {@link Notification notifications}.
+ *
+ * <p>A repository is responsible for handling CRUD and data parsing operations around these
+ * objects, in whatever scheme necessary.
+ *
+ * <p>Production implementations are expected to be thread safe, even across a network.
+ */
 public interface NotificationRepository {
 
     /**
-     * Retrieves the oldest {@code maxEvents} {@link Notification notifications}, oldest first.
+     * Retrieves the oldest {@code maxNotifications} {@link Notification notifications}, oldest
+     * first.
      *
      * <p>Once retrieved, a notification should not be retrieved again, atomically. That is, many
-     * threads looking up notifications at the same time should all et a unique non-overlapping sample
-     * of the oldest notifications. Subsequent calls should always return a unique set.
+     * threads looking up notifications at the same time should all et a unique non-overlapping
+     * sample of the oldest notifications. Subsequent calls should always return a unique set.
```

Comment:
Actually I just realized that weighted randomness doesn't matter too much if you don't grab much more than you will try to lock (in other words, if you are trying to lock everything, then there is no risk of something not being processed of course). I might implement this sooner.

For doc events it's a little different because we are grabbing a larger batch than what we will actually process. You had a suggestion below that would change that though, so I'll think about that.